### ISSUE TITLE
cdrizzle.tdriz crashes with segmentation fault and when it "works" the drizzled images have 'fillval' only

### ISSUE THREAD

**Issue by mcara** [spacetelescope/drizzle#1]
I created a simple test of drizzling a very simple image onto a larger image. The coordinate transformation is an identity transformation so the input image should be in a quadrant of the output image. The code that crashes `tdriz`:

```python
>>> import numpy as np
>>> from drizzle import cdrizzle
>>> inmask=np.zeros((100,100),dtype=np.float32)
>>> inmask[22:25,33:49]=1
>>> inwht=np.ones_like(inmask,dtype=np.float32)
>>> outsci=np.zeros((256,256),dtype=np.float32)
>>> outwht=np.zeros((256,256),dtype=np.float32)
>>> outctx=np.zeros((256,256),dtype=np.int32)
>>> in_y, in_x = np.indices((100,100), dtype=np.float)
>>> pixmap = np.dstack((in_y,in_x))
>>> cdrizzle.tdriz(inmask,inwht,pixmap,outsci,outwht,outctx,
... 1,0,0,0,0,1,1,'square','cps',1,1,'-1')
('Callable C-based DRIZZLE Version 0.9 (10th May 2014)', 65536, 156)

>>> # second try with identical parameters:
>>> import numpy as np
>>> from drizzle import cdrizzle
>>> inmask=np.zeros((100,100),dtype=np.float32)
>>> inmask[22:25,33:49]=1
>>> inwht=np.ones_like(inmask,dtype=np.float32)
>>> outsci=np.zeros((256,256),dtype=np.float32)
>>> outwht=np.zeros((256,256),dtype=np.float32)
>>> outctx=np.zeros((256,256),dtype=np.int32)
>>> in_y, in_x = np.indices((100,100), dtype=np.float64)
>>> pixmap = np.dstack((in_y,in_x))
>>> cdrizzle.tdriz(inmask,inwht,pixmap,outsci,outwht,outctx,
... 1,0,0,0,0,1,1,'square','cps',1,1,'-1')
Segmentation fault
```
In the first run, even though `tdriz` did not crash, it did "miss" all of the data points (which should not happen for an identity coordinate transformation):
```python
>>> np.any(outsci>0)
False
>>> np.sum(outsci>0)
0
```
The old code from DrizzlePac works as expected:
```python
>>> from drizzlepac import cdriz
>>> import numpy as np
>>> inmask=np.zeros((100,100),dtype=np.float32)
>>> inmask[22:25,33:49]=1
>>> inwht=np.ones_like(inmask,dtype=np.float32)
>>> outsci=np.zeros((256,256),dtype=np.float32)
>>> outwht=np.zeros((256,256),dtype=np.float32)
>>> outctx=np.zeros((256,256),dtype=np.int32)
>>> cdriz.tdriz(inmask,inwht,outsci,outwht,outctx,0,0,1,1,inmask.shape[0],
... 1.0,1.0,1.0,'center',1.0,'square','cps',1,1,'-1',0,0,1,lambda x,y:[np.asarray(x),np.asarray(y)])
('Callable C-based DRIZZLE Version 0.8 (20th May 2009)', 0, 0) # <-- nothing missed!
>>> np.any(outsci>0)
True
>>> np.sum(outsci>0)
48
```

**Comment by mdboom** [comment]
Can you get a gdb traceback?

**Comment by mcara** [comment]
Did you try to run my example (a couple of times, maybe) and it works fine for you?

**Comment by bernie-simon** [comment]
On Jun 29, 2015, at 4:22 PM, Mihai Cara <notifications@github.com> wrote:

> Did you try to run my example (a couple of times, maybe) and it works fine for you?

I haven�t been able to produce the segmentation violation yet. Right now I am trying to figure out why it�s reporting all the pixels as missed.

**Comment by mdboom** [comment]
No, I haven't tried to run it.  But the most helpful bit of information from a segfault is usually the backtrace.

```
gdb `which python`
> run
... do whatever it takes to reproduce...
> bt
```

**Comment by mcara** [comment]
```python
>>> cdrizzle.tdriz(inmask,inwht,pixmap,outsci,outwht,outctx,1,0,0,0,0,1,1,'square','cps',1,1,'-1') # crashes with segfault

Program received signal EXC_BAD_ACCESS, Could not access memory.
Reason: KERN_INVALID_ADDRESS at address: 0x0000000103b00000
0x00000001039981d4 in set_pixel ()
(gdb) 
```

**Comment by bernie-simon** [comment]
On Jun 29, 2015, at 4:45 PM, Mihai Cara <notifications@github.com> wrote:

> >>> cdrizzle.tdriz(inmask,inwht,pixmap,outsci,outwht,outctx,1,0,0,0,0,1,1,'square','cps',1,1,'-1') # crashes with segfault
> 
> 
> Program received signal EXC_BAD_ACCESS, Could 
> not
>  access memory.
> Reason: KERN_INVALID_ADDRESS at address: 
> 0x
> 0000000103b00000
> 
> 0x00000001039981d4 in
>  set_pixel ()
> (gdb) 
> 
> �
> Reply to this email directly or view it on GitHub.
> 

Which probably means it�s getting bad pixel coordinates.

**Comment by mcara** [comment]
_"Which probably means it�s getting bad pixel coordinates."_ From who?

**Comment by bernie-simon** [comment]
On Jun 30, 2015, at 10:57 AM, Mihai Cara <notifications@github.com> wrote:

> From who?

I�ll let you know when I identify the problem.

**Comment by mdboom** [comment]
drizzle used to have an option to build the C code in "debug mode", which turns on range checking asserts all over the code (and also makes it run significantly slower).  It looks like when the conversion from raw distutils to astropy was done, however, this flag was removed.  However, setting the environment variable `CFLAGS` to `-DDEBUG -UNDEBUG -Og -g` and then doing a full rebuild may be enough (haven't verified that).

**Comment by bernie-simon** [comment]
Range checking was removed because the way arrays are accessed in drizzle was changed, In the past array dimensions were passed as separate parameters, now the drizzle code passes around  a pointer to the numpy data structure and uses numpy's macros to access the individual pixels.

As part of the fix I've added bounds checking back. Would have added it earlier, but it took me a while to figure out how to interface with drizzle's error handling in a clean way. Now it throws a Python error if there is a an out of bounds array access.

The segmentation violation was caused by an out of bounds access when setting the fill value. I changed the for loop limits to fix that.

The code was missing every point because the function computing the overlap area between the input and output pixel was always negative. I changed the formula to return a positive result in this case and take the absolute value of the area before returning it as an additional precaution.

**Comment by mdboom** [comment]
On github, it is best practice to submit a PR so the code can be reviewed before it is merged.  I have made some comments on the associated commit.

> As part of the fix I've added bounds checking back. Would have added it earlier, but it took me a while to figure out how to interface with drizzle's error handling in a clean way. Now it throws a Python error if there is a an out of bounds array access.

When the Fortran to C conversion was first done, the range checking ended up dominating the run time by a significant fraction.  Importantly, though, they were implemented as C asserts, so when not compiled in debug mode, they had no impact on runtime performance.  There seems to be no need to do range checking all the time -- any out of bounds error would be the fault of a bug in the C code, not anything the Python programmer may have passed in, so it seems unimportant to do all this checking to make Python exceptions.
